### PR TITLE
journal: Deduplicate entry items before they are stored in the entry object

### PR DESCRIPTION
**Pull request by daandemeyer**
If the iovec contains the same data more than once, we'll end up with
duplicate offsets in the items array. Let's make sure we remove any
duplicates before we store the items in an entry object.

Fixes entry array not sorted --verify errors.

Includes #21306, #21462 and #21463 (only the last commit is new)

### CODE REVIEW

**Comment by DaanDeMeyer** [comment]
test-time-util failure, dissect failure and the two expected timeouts

**Comment by poettering** [comment]
the third patch looks good. 2nd patch i'll review in the other pr

File: src/libsystemd/sd-journal/journal-file.c [hunk 1]
```diff
@@ -2078,6 +2078,19 @@ static int entry_item_cmp(const EntryItem *a, const EntryItem *b) {
         return CMP(le64toh(a->object_offset), le64toh(b->object_offset));
 }
 
+static size_t remove_duplicate_entry_items(EntryItem *items, size_t n) {
+        size_t j = 1;
+
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-file.c:2085]: maybe add comment here that this relies on the fact that the array is already sorted

File: src/libsystemd/sd-journal/journal-file.c [hunk 2]
```diff
@@ -2078,6 +2078,19 @@ static int entry_item_cmp(const EntryItem *a, const EntryItem *b) {
         return CMP(le64toh(a->object_offset), le64toh(b->object_offset));
 }
 
+static size_t remove_duplicate_entry_items(EntryItem *items, size_t n) {
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-file.c]: doesn't matter, but if you'd replace `*items` by `items[]` here, things would be a tiny bit clearer, even though of course it's entirely equivalent

File: src/libsystemd/sd-journal/journal-file.c [hunk 3]
```diff
@@ -2078,6 +2078,19 @@ static int entry_item_cmp(const EntryItem *a, const EntryItem *b) {
...
+        size_t j = 1;
+
+        if (n == 0)
+                return 0;
```
Comment (poettering) [inline src/libsystemd/sd-journal/journal-file.c]: doesn't matter, but you could write this as:

```c
if (n <= 1)
        return n;
```

i.e. zero length and single item arrays never have duplicates...

**Comment by poettering** [comment]
Before merged should probably be rebase on the other two PRs once they are merged

**Comment by DaanDeMeyer** [comment]
Rebased and addressed comments

**Pull request merged**